### PR TITLE
Use git attributes to determine generated and vendored status for language stats and diffs

### LINKED ISSUE #12653: [Feature] Ability to exclude files from diffs
Github uses Linguist for code analysis in its repositories, and resulting from that introduced a nifty feature in its .gitattributes definitions: `linguist-generated`. This attribute can be used to exclude specific file types from diffs, code reviews, and hide it from showing in the file listing.
I've found that to be invaluable to working with Unity3D in order to keep .meta and .prefab files from cluttering code reviews. It would be cool to have a similar feature in Gitea

---

[Bountysource](https://www.bountysource.com/issues/92731348-feature-ability-to-exclude-files-from-diffs)

### LINKED ISSUE #14786: [Feature Request]: Ignore directories from language statistics bar
- Gitea version (or commit ref): 1.13.0
- Git version: N/A
- Operating system: NixOS/ Linux (systemd)

- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io: N/A
- Log gist: N/A

## Description

The `linguist` project used by GitHub (and GitLab) supports `.gitattributes` files in the repository tree to exclude certain directories or files.  This is very useful when vendoring large code-bases.

I operate a monorepo for all my code, which contains subtrees for many of my project dependencies. Because of this the language bar is not representative of the languages I actually write.

It would be great if there was some way of telling gitea to ignore certain directories for the purpose of language statistics. I had a look at the code, but I really don't know Go and couldn't figure out how to implement it.

It doesn't neccessarily have to be the same `.gitattributes` format that `linguist` uses (it would be cool though!). It would already be enough if there was a list of "ignore paths" in the project settings screen.

Anyway, let me know what you think.

### PR DESCRIPTION
**Pull request by zeripath**
Replaces #16262
Replaces #16250
Replaces #14833

This PR first implements a `git check-attr` pipe reader - using `git check-attr --stdin -z --cached` - taking account of the change in the output format in git 1.8.5 and creates a helper function to read a tree into a temporary index file for that pipe reader.

It then wires this in to the language stats helper and into the git diff generation.

Files which are marked generated will be folded by default.

Fixes #14786
Fixes #12653

### CODE REVIEW

File: templates/repo/diff/box.tmpl [hunk 1]
```diff
@@ -65,6 +69,12 @@
 							{{end}}
 						</div>
 						<span class="file mono">{{if $file.IsRenamed}}{{$file.OldName}} &rarr; {{end}}{{$file.Name}}{{if .IsLFSFile}} ({{$.i18n.Tr "repo.stored_lfs"}}){{end}}</span>
+						{{if $file.IsGenerated}}
+							<span class="ui label ml-3 mr-3">{{$.i18n.Tr "repo.diff.generated"}}</span>
+						{{end}}
+						{{if $file.IsVendored}}
```
Comment (delvh) [inline templates/repo/diff/box.tmpl:75]: If I see that correctly, couldn't that potentially result in a file where both the `isGenerated` and the `isVendored` are displayed, assuming a vendored and generated file would exist?
Would it be better to make this an `else if` then or can this case be ignored?

Comment (zeripath) [reply templates/repo/diff/box.tmpl:75]: Yes vendored and generated could be both displayed. They appear to be orthogonal statuses and we calculate both so I don't see a good reason not to display both.

File: modules/git/repo_attribute_test.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
Comment (lunny) [inline modules/git/repo_attribute_test.go]: 2021

File: modules/git/repo_language_stats_gogit.go [hunk 3]
```diff
@@ -2,16 +2,19 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
+//go:build gogit
```
Comment (lunny) [inline modules/git/repo_language_stats_gogit.go:5]: This will require go1.17?

File: templates/repo/diff/box.tmpl (hunk 1 above)
Comment (delvh) [reply templates/repo/diff/box.tmpl:75]: Okay, yeah, I can see that.

**Comment by zeripath** [comment]
@silverwind would you like to re-review?

File: modules/git/repo_language_stats_gogit.go [hunk 4]
```diff
@@ -42,9 +44,73 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
...
+				if generated, has := attrs["linguist-generated"]; has {
+					if generated == "set" || generated == "true" {
+						return nil
+					}
+					notGenerated = generated == "false"
+				}
+				if language, has := attrs["linguist-language"]; has && language != "unspecified" && language != "" {
+					// group languages, such as Pug -> HTML; SCSS -> CSS
+					group := enry.GetLanguageGroup(language)
+					if group != "" {
```
Comment (lafriks) [inline modules/git/repo_language_stats_gogit.go]: We mostly use `len(x) == 0`

File: modules/git/repo_language_stats_nogogit.go [hunk 5]
```diff
@@ -62,13 +63,78 @@ func (repo *Repository) GetLanguageStats(commitID string) (map[string]int64, err
...
+				if generated, has := attrs["linguist-generated"]; has {
+					if generated == "set" || generated == "true" {
+						continue
+					}
+					notGenerated = generated == "false"
+				}
+				if language, has := attrs["linguist-language"]; has && language != "unspecified" && language != "" {
+					// group languages, such as Pug -> HTML; SCSS -> CSS
+					group := enry.GetLanguageGroup(language)
+					if group != "" {
```
Comment (lafriks) [inline modules/git/repo_language_stats_nogogit.go]: Same here

File: modules/git/repo_language_stats_gogit.go (hunk 4 above)
Comment (zeripath) [reply modules/git/repo_language_stats_gogit.go]:
```suggestion
					if len(group) == 0 {
```

File: modules/git/repo_language_stats_nogogit.go (hunk 5 above)
Comment (zeripath) [reply modules/git/repo_language_stats_nogogit.go]:
```suggestion
					if len(group) == 0 {
```

**Review by silverwind — APPROVED** [review]

**Comment by delvh** [comment]
> Screenshot from 2021-08-27 09-20-38
> 
> If you click the `>` the diff will be shown.

Does that mean that the diff is already precomputed even if the file is most likely never seen?
I would rather suggest lazy-loading the diff the first time it is being requested (i.e. when expanding it, or explicitly using a "Load Diff" button inside it) because especially for repos that have lots of generated (and vendored, if you decide to collapse it again by default) files can then be handled almost instantly, resulting in a way better initial performance. As a result, Gitea would be even more responsive under normal circumstances.

By the way, apart from the `.gitignore`, I would also say that vendored files should be lazy-loaded by default.

**Comment by zeripath** [comment]
> But, to be fair, I am not exactly sure whether that fits inside the scope of this branch.
> I don't know how many fundamentals for lazy-loading are already present, so this might potentially be a whole separate issue.

It doesn't fit within the scope of this PR, no. That would seriously and unnecessarily complicate 

I need the ideas in #16829 and some the extension work blocked by that for lazy loading to become possible - but I'm blocked til both of these PRs are merged.

**Review by lunny — APPROVED** [review]

**Pull request merged**